### PR TITLE
Advent 2017: go execution tracer (Dec 23th)

### PR DESCRIPTION
**Pull request by andrestc**
Feedback welcome. I plan to do some reviewing myself but mostly for English mistakes and some rewording. 

Would like to get some feedback as to whether I should add more content on a particular topic.

### CODE REVIEW

File: content/advent-2017/go-execution-tracer.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+		ch <- 42
+	}()
+	// read from channel
+	<-ch
+}
+```
+
+This example creates an unbuffered channel and initializes a goroutine that will send the number 42 over this channel. The main goroutine blocks until the other goroutines send a value over the channel.
+
+Running this code with `go run main.go 2> trace.out`, sends the tracing output to the file `trace.out`, which can then be read with: `go tool trace trace.out`.
```
Comment (nictuku) [inline content/advent-2017/go-execution-tracer.md]: s/sends/it sends/

File: content/advent-2017/go-execution-tracer.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+	Provides timing for scheduler level information showing where time is most spent scheduling.
+
+## View Trace
+
+Clicking on the "View trace" linking, one is presented with a screen full of information about the whole program execution. 
+
+> Press "?" to get a list of available shortcuts to help navigating the trace.
+
+The following image highlights the most important parts and each section is describe below:
```
Comment (nictuku) [inline content/advent-2017/go-execution-tracer.md]: s/describe/described/

File: content/advent-2017/go-execution-tracer.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+![View trace web](/postimages/advent-2017/go-execution-tracer/trace-web.png)
+
+This particular screenshot shows a GC run that starts at between 1169ms and 1170ms and ends right after 1174ms. During this time, an OS thread (PROC 1) ran a goroutine dedicated to the GC while other goroutines assisted in some GC phases (these are displayed on lines bellow the goroutine and are read MARK ASSIST). By the end of the screenshot, we can see that most of the allocated memory was freed by the GC.
+
+Another particular useful information is the number of goroutines that are in "Runnable" state (13 on the selected time): if this number becomes large over time this can indicate that we need more CPUs to handle the load.
+
+# Conclusions
+
+The tracer is a powerful tool for debugging concurrency issues, e.g, contentions and logical races. But it does not solves all problems: it is not the best tool available to track down what piece of code is spending most CPU time or allocations. The `go tool pprof` is better suited for these use cases.
```
Comment (nictuku) [inline content/advent-2017/go-execution-tracer.md]: s/it does not solves/it does not solve/

File: content/advent-2017/go-execution-tracer.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+The tool really shines when you want to understand the behavior of a program over time and to know what each goroutine is doing when NOT running. Collecting traces may have some overhead and can generate a high amount of data to be inspected.
+
+Unfortunately, there is not many documentation available so some experimentation is needed to try and understand what the tracer is showing. This is also an [opportunity for contributions](https://github.com/golang/go/issues/16526) to the official documentation and to the community in general (e.g blog posts).
```
Comment (nictuku) [inline content/advent-2017/go-execution-tracer.md]: I don't know a better way to say it, but "there is not many documentation available" doesn't seem right. Maybe "documentation is lacking" or so.

File: content/advent-2017/go-execution-tracer.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+title = "Go execution tracer"
+linktitle = "Go execution tracer"
+date = 2017-12-23T20:00:00Z
+series = ["Advent 2017"]
+
++++
+
+# Introduction
+
+Ever wondered how are your goroutines being scheduled by the go runtime? Ever tried to understand why adding concurrency to your program has not given it better performance? The go execution tracer can help answer this and other questions to help you diagnose performance issues, e.g, latency, contention and poor parallelization.
```
Comment (Pisush) [inline content/advent-2017/go-execution-tracer.md]: `this`=`these`?

File: content/advent-2017/go-execution-tracer.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+1. Creation, start and end of goroutines
+2. Events that block/unblock goroutines (syscalls, channels, locks)
+3. Network I/O related events
+4. Syscalls
+5. Garbage collection
+
+All this data is collected by the tracer without any kind of aggregation or sampling. In some busy applications this may result in a large file that can be analyzed afterwards by the `go tool trace` command. 
+
+Go already had the pprof memory and CPU profiler, before the introduction of the execution tracer, so why was it added to the official toolchain? While the CPU profiler does a nice job to telling you what function is spending most CPU time, it does not help you figure out what is preventing a goroutine from running or how are the goroutines being scheduled on the available OS threads. Thats precisely where the tracer really shines. The tracer [design doc](https://docs.google.com/document/u/1/d/1FP5apqzBgr7ahCCgFO-yoVhk4YZrNIDNf9RybngBc14/pub) does a pretty good job explaining the motivations behind the tracer and how it was designed to work.
```
Comment (Pisush) [inline content/advent-2017/go-execution-tracer.md]: `Thats precisely = `That's precisely`

**Comment by andrestc** [comment]
Thanks for reviewing, @nictuku and @Pisush.

**Comment by object88** [comment]
Hello @andrestc !  I'm a fan of using pprof myself, but it's very easy to accidentally open your application to some vulnerabilities.  There is a [blog post](http://mmcloughlin.com/posts/your-pprof-is-showing) or [two](https://www.farsightsecurity.com/2016/10/28/cmikk-go-remote-profiling/) on the subject, and I'd encourage you to share these with your readers, or update your sample code to reduce the vulnerability.  Thanks!

**Comment by andrestc** [comment]
@object88 I've updated the text adding a note and link to one of the articles. Thanks!

**Pull request merged**